File: RentalReal.java
package videoprokat;

import java.util.ArrayList;
import java.util.List;

/**
 * Реализация класса Rental.
 * Содержит конкретную реализацию всех методов абстрактного класса Rental.
 */
public class RentalReal extends Rental {

    /**
     * Конструктор по умолчанию.
     */
    public RentalReal() {
        this.items = new ArrayList<>();
    }

    /**
     * Конструктор с параметрами.
     */
    public RentalReal(int rentalId, Client client, List<VideoCarrier> items,
                      String rentalDate, String plannedReturnDate) {
        super(rentalId, client, new ArrayList<>(items), rentalDate, plannedReturnDate);
        this.items = new ArrayList<>(items);
    }

    @Override
    public int getRentalId() {
        return rentalId;
    }

    @Override
    public Client getClient() {
        return client;
    }

    @Override
    public List<VideoCarrier> getItems() {
        return items;
    }

    @Override
    public String getRentalDate() {
        return rentalDate;
    }

    @Override
    public String getPlannedReturnDate() {
        return plannedReturnDate;
    }

    @Override
    public double getDepositAmount() {
        return depositAmount;
    }

    @Override
    public double getRentalCost() {
        return rentalCost;
    }

    @Override
    public String getStatus() {
        return status;
    }

    @Override
    public void setCalculatedAmounts(FinancialCalculator calculator, int days) {
        if (calculator != null) {
            depositAmount = calculator.calculateDeposit(items);
            rentalCost = calculator.calculateRentalCost(items, days);
        }
    }

    @Override
    public double closeRental(double overdueFine) {
        status = "closed";
        return rentalCost + overdueFine;
    }

    /**
     * Деконструктор (финализатор).
     * Вызывается перед удалением объекта сборщиком мусора.
     */
    @Override
    protected void finalize() throws Throwable {
        try {
            if (items != null) {
                items.clear();
            }
        } finally {
            super.finalize();
        }
    }
}

File: cpp/RentalReal.cpp
#include "RentalReal.hpp"
#include "Client.hpp"
#include "VideoCarrier.hpp"
#include "FinancialCalculator.hpp"

RentalReal::RentalReal()
{
    items.clear();
}

RentalReal::RentalReal(int rentalId, Client* client, const std::vector<VideoCarrier*>& items,
                       const std::string& rentalDate, const std::string& plannedReturnDate)
    : Rental(rentalId, client, items, rentalDate, plannedReturnDate)
{
}

RentalReal::~RentalReal()
{
    // В C++ не удаляем items, так как они принадлежат другим объектам
    items.clear();
}

int RentalReal::getRentalId()
{
    return rentalId;
}

Client* RentalReal::getClient()
{
    return client;
}

std::vector<VideoCarrier*> RentalReal::getItems()
{
    return items;
}

std::string RentalReal::getRentalDate()
{
    return rentalDate;
}

std::string RentalReal::getPlannedReturnDate()
{
    return plannedReturnDate;
}

double RentalReal::getDepositAmount()
{
    return depositAmount;
}

double RentalReal::getRentalCost()
{
    return rentalCost;
}

double RentalReal::getTotalCost()
{
    return depositAmount + rentalCost;
}

std::string RentalReal::getStatus()
{
    return status;
}

void RentalReal::setCalculatedAmounts(FinancialCalculator* calculator, int days)
{
    if (calculator != nullptr) {
        depositAmount = calculator->calculateDeposit(items);
        rentalCost = calculator->calculateRentalCost(items, days);
    }
}

double RentalReal::closeRental(double overdueFine)
{
    status = "closed";
    return rentalCost + overdueFine;
}

